external/error_report/error_report.c

```diff
@@ -0,0 +1,205 @@
+/****************************************************************************
+ *
+ * Copyright 2018 Samsung Electronics All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+#include <tinyara/config.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <time.h>
+#include <pthread.h>
+#include <unistd.h>
+#include <libgen.h>
+#include <sys/socket.h>
+#include <arpa/inet.h>
+#include <sched.h>
+#include <error_report/error_report.h>
+#include "sched/sched.h"
+
+#define ERR_IP_ADRR_LEN 16
+
+#ifdef CONFIG_ERROR_REPORT_DEBUG
+#define rt_errdebug dbg
+#else
+#define rt_errdebug(...)
+#endif
+
+nw_error_data_t g_error_report[CONFIG_ERROR_REPORT_NENTRIES];
+typedef enum {
+	ERRSTATE_UNINITIALIZED = -1,
+	ERRSTATE_INITIALIZED,
+	ERRSTATE_MEMUNDERFLOW,
+	ERRSTATE_BUFFER_EXCEEDED,
+} err_fsm_t;
+
+typedef struct {
+	uint16_t rear;
+	uint16_t front;
+	int8_t q_pending;
+	pthread_mutex_t err_mutex;
+	err_fsm_t fsm;
+} g_error_info_t;
+
+static g_error_info_t g_err_info;
+
+static unsigned long prv_fetch_taskaddr(int pid)
+{
+	int i;
+
+	for (i = 0; i < CONFIG_MAX_TASKS; i++) {
+		if ((g_pidhash[i].tcb)->pid == pid) {
+			entry_t e;
+			e = g_pidhash[i].tcb->entry;
+			return (unsigned long)e.pthread;
```

Comment:
How do you know it is a pthread, not task?

Reply:
Point noted, and will address the changes. Thanks.